packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -47,13 +48,31 @@ export async function tryHotswapDeployment(
 
   const currentTemplate = await loadCurrentTemplateWithNestedStacks(stackArtifact, sdk);
   const stackChanges = cfn_diff.diffTemplate(currentTemplate.deployedTemplate, stackArtifact.template);
-  const hotswappableChanges = await findAllHotswappableChanges(
+  const { hotswappableChanges, nonHotswappableChanges, metadataChanged } = await classifyResourceChanges(
     stackChanges, evaluateCfnTemplate, sdk, currentTemplate.nestedStackNames,
   );
 
-  if (!hotswappableChanges) {
-    // this means there were changes to the template that cannot be short-circuited
-    return undefined;
+  if (nonHotswappableChanges.length > 0) {
+    logNonHotswappableChanges(nonHotswappableChanges);
+  }
+
+  // preserve classic hotswap behavior
+  if (hotswapMode === HotswapMode.CLASSIC) {
+    // The only change detected was to CDK::Metadata, so return noOp
+    if (hotswappableChanges.length === 0 && nonHotswappableChanges.length === 0 && metadataChanged) {
+      return { noOp: true, stackArn: cloudFormationStack.stackId, outputs: cloudFormationStack.outputs };
+    }
+
+    // the number of hotswappable resources is less than the number of changes
+    // this means that at least one change was deemed not hotswappable by every detector function,
+    // ie they all returned { hotswappableChanges: [], ... }
+    if (hotswappableChanges.length < Object.keys(getStackResourceDifferences(stackChanges)).length) {
```

Comment:
you're right, the ones that the detectors had no opinion on should be reported as non-hotswappable changes.

I think we did this to prevent having to do that mapping all in one place, since it got more complicated as we added more resources; `Version`s, `Alias`es, and `Function`s all go to Lambda, for example, and IAM Policies go to s3-bucket-deployment. In the future, other resources might also need to check IAM resource changes. We figured it would be better to keep it spread across files than in one central spot.